src/Pim/Bundle/EnrichBundle/Connector/Reader/MassEdit/FilteredProductModelReader.php

```
@@ -0,0 +1,206 @@
+        ChannelRepositoryInterface $channelRepository,
+        CompletenessManager $completenessManager,
+        MetricConverter $metricConverter,
+        $generateCompleteness
+    ) {
+        $this->pqbFactory = $pqbFactory;
+        $this->channelRepository = $channelRepository;
+        $this->completenessManager = $completenessManager;
+        $this->metricConverter = $metricConverter;
+        $this->generateCompleteness = (bool) $generateCompleteness;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function initialize(): void
+    {
+        $channel = $this->getConfiguredChannel();
+        if (null !== $channel && $this->generateCompleteness) {
+            $this->completenessManager->generateMissingForChannel($channel);
+        }
+
+        $filters = $this->getConfiguredFilters();
+        $this->productsAndProductModels = $this->getProductsCursor($filters, $channel);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function read(): ?ProductModelInterface
+    {
+        $product = null;
+        $product = $this->getNextProduct();
+
+        if (null !== $product) {
+            $channel = $this->getConfiguredChannel();
+            if (null !== $channel) {
+                $this->metricConverter->convert($product, $channel);
+            }
+        }
+
+        return $product;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setStepExecution(StepExecution $stepExecution): void
+    {
+        $this->stepExecution = $stepExecution;
+    }
+
+    /**
+     * Returns the configured channel from the parameters.
+     * If no channel is specified, returns null.
+     *
+     * @throws ObjectNotFoundException
+     *
+     * @return ChannelInterface|null
+     */
+    private function getConfiguredChannel(): ?ChannelInterface
+    {
+        $parameters = $this->stepExecution->getJobParameters();
+        if (!isset($parameters->get('filters')['structure']['scope'])) {
```

Why could the channel be empty? The grid is display depending on a channel. You should throw an exception here